src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/TransportCreation/TransportCreationHandler.cs

```diff
@@ -0,0 +1,215 @@
+            foreach (var ipAddress in ipAddresses)
+            {
+                ipEndpoint.Address = ipAddress;
+                try
+                {
+                    socket = await OpenSocket(ipEndpoint, isAsync, ct).ConfigureAwait(false);
+                    break;
+                }
+                catch(Exception e)
+                {
+                    socketOpenExceptions.Add(e);
+                }
+            }
+
+            // If no socket succeeded, throw
+            if (socket is null)
+            {
+                // If there are any socket exceptions in the collected exceptions, throw the first
+                // one. If there are not, collect all exceptions and throw them as an aggregate.
+                foreach (var exception in socketOpenExceptions)
+                {
+                    if (exception is SocketException)
+                    {
+                        throw exception;
+                    }
+                }
+
+                throw new AggregateException(socketOpenExceptions);
+            }
+
+            // Create the stream for the socket
+            return new NetworkStream(socket);
+        }
+
+        private async ValueTask<Socket> OpenSocket(IPEndPoint ipEndPoint, bool isAsync, CancellationToken ct)
+        {
+            ct.ThrowIfCancellationRequested();
+
+            var socket = new Socket(ipEndPoint.AddressFamily, SocketType.Stream, ProtocolType.Tcp) { Blocking = false };
+
+            // Enable keep-alive
+            socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.KeepAlive, true);
+            socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveInterval, KeepAliveIntervalSeconds);
+            socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveTime, KeepAliveTimeSeconds);
+
+            try
+            {
+                if (isAsync)
+                {
+                    #if NET6_0_OR_GREATER
+                    await socket.ConnectAsync(ipEndPoint, ct).ConfigureAwait(false);
+                    #else
+                    // @TODO: Only real way to cancel this is to register a cancellation token event and dispose of the socket.
+                    await new TaskFactory(ct).FromAsync(socket.BeginConnect, socket.EndConnect, ipEndPoint, null)
+                        .ConfigureAwait(false);
+                    #endif
+                }
+                else
+                {
+                    OpenSocketSync(socket, ipEndPoint, ct);
+                }
+            }
+            catch (Exception)
+            {
+                socket.Dispose();
+                throw;
+            }
+
+            // Connection is established
+            socket.Blocking = true;
+            socket.NoDelay = true;
+
+            return socket;
+        }
+
+        private void OpenSocketSync(Socket socket, IPEndPoint ipEndPoint, CancellationToken ct)
```

Comment:
There are certain improvements that can be made for sync paths. Let me know what you think. 

The caller of this code is iterating over the IP addresses and calling `OpenSocketSync`

`OpenSocketSync` will block till it can connect or fail. This means that serially connecting to the sockets would take up a bunch of time before the good IP address gets its turn.

For sync opens, you could probably provide a list of IP addresses and create a socket for each IP with `Blocking = false` 
After this, call the socket.Connect for each of these sockets. Hence `socket.Connect` will execute immediately with an error message. 
If the `e.SocketErrorCode` is WouldBlock, then don't handle the exception, because this means that the socket may connect later. 

At the end of the loop call `Socket.Select` on all the sockets to see which one is open, and use that particular socket. This allows you to quickly schedule connections to each socket of the IP addresses resolved and then wait for the first one that is connected.

Reply:
You may not make this change now, but this is definitely very important to MultisubnetFailover (which is not the scope of this PR) where the docs say that an endpoint Could resolve to 64 IP addresses.

Reply:
Yes, this is basically how MSF will work, however that is for a later PR.